src/AutoRest.CSharp/Common/Input/CadlInputConstantConverter.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using AutoRest.CSharp.Generation.Types;
+using AutoRest.CSharp.Output.Models.Types;
+using static AutoRest.CSharp.Input.Configuration;
+
+namespace AutoRest.CSharp.Common.Input
+{
+    internal class CadlInputConstantConverter: JsonConverter<InputConstant>
+    {
+        private readonly CadlReferenceHandler _referenceHandler;
+
+        public CadlInputConstantConverter(CadlReferenceHandler referenceHandler)
+        {
+            _referenceHandler = referenceHandler;
+        }
+
+        public override InputConstant Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            => reader.ReadReferenceAndResolve<InputConstant>(_referenceHandler.CurrentResolver) ?? CreateInputConstant(ref reader, null, null, options, _referenceHandler.CurrentResolver);
+
+        public override void Write(Utf8JsonWriter writer, InputConstant value, JsonSerializerOptions options)
+            => throw new NotSupportedException("Writing not supported");
+
+        public static InputConstant CreateInputConstant(ref Utf8JsonReader reader, string? id, string? name, JsonSerializerOptions options, ReferenceResolver resolver)
+        {
+            var isFirstProperty = id == null;
+            Object? value = null;
+            InputType? type = null;
+            while (reader.TokenType != JsonTokenType.EndObject)
+            {
+                var isKnownProperty = reader.TryReadReferenceId(ref isFirstProperty, ref id)
```

Comment:
Since there can be only two properties (Type and Value) and we require type to go after value, I think we can enforce strict order here:
```cs
reader.TryReadReferenceId(ref isFirstProperty, ref id);
if (!TryReadWithConverter(nameof(InputConstant.Type), options, ref type)) 
{
    throw 
}
ReadConstantValue(ref reader, nameof(InputConstant.Value), options, type, ref value);
```